### PR TITLE
boards: arm: lpcxpresso51u68: Add LPCXpresso51U68 board definition

### PR DESCRIPTION
**Pull request by Navin-Sankar**
Add SoC and board definition for the NXP LPCXpresso51U68 development
board.

Added adc driver support for LPC51 series soc.

Signed-off-by: Navin Sankar Velliangiri <navin@linumiz.com>

### CODE REVIEW

File: drivers/gpio/gpio_mcux_lpc.c [hunk 1]
```diff
@@ -230,11 +230,19 @@ static uint32_t attach_pin_to_isr(uint32_t port, uint32_t pin, uint32_t isr_no)
 	/* Code asumes PIN_INT values are grouped [0..3] and [4..7].
 	 * This scenario is true in LPC54xxx/LPC55xxx.
 	 */
+#ifdef CONFIG_BOARD_LPCXPRESSO51U68
```
Comment (galak) [inline drivers/gpio/gpio_mcux_lpc.c:233]: We need to find a better way to handle this, we shouldn't introduce board specific ifdef's in this file.

Comment (galak) [reply drivers/gpio/gpio_mcux_lpc.c:233]: Also, does this make sense on LPC51U68 in that it seems like only has PINT0/1/2/3 so the else case doesn't make any sense.

**Review by galak — CHANGES_REQUESTED** [review]
Need to cleanup how we handle the PIN_INT differences for LPC51U68.

**Comment by Navin-Sankar** [comment]
@galak can this tagged for 2.7?

File: drivers/gpio/gpio_mcux_lpc.c [hunk 2]
```diff
@@ -51,8 +51,13 @@ struct gpio_mcux_lpc_data {
 	sys_slist_t callbacks;
 	/* pin association with PINT id */
 	pint_pin_int_t pint_id[32];
+#ifdef CONFIG_SOC_SERIES_LPC51U68
 	/* ISR allocated in device tree to this port */
-	uint32_t isr_list[8];
+	uint32_t isr_list[4];
```
Comment (dleach02) [inline drivers/gpio/gpio_mcux_lpc.c:56]: Instead of having conditional code based on the platform, there is a define in the SOC header file for the count here that would be cleaner:

INPUTMUX_PINTSEL_COUNT

Comment (Navin-Sankar) [reply drivers/gpio/gpio_mcux_lpc.c]: @dleach02 fixed.

File: modules/Kconfig.mcux [hunk 3]
```diff
@@ -25,6 +25,11 @@ config HAS_MCUX_LPADC
 	help
 	  Set if the LPADC module is present in the SoC.
 
+config HAS_LPC_ADC
+	bool
+	help
+	  Set if the LPC ADC module is present in the SoC.
+
```
Comment (MaureenHelm) [inline modules/Kconfig.mcux:32]: This change should be moved to the commit `drivers: adc: added support adc drive for lpc51u68` or a new commit

File: soc/arm/nxp_lpc/lpc51u68/Kconfig.defconfig.series [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+config SOC_FLASH_LPC
+	default y
+	depends on FLASH
+
+config PWM_MCUX_SCTIMER
+	default y
+	depends on PWM
+
+config HAS_LPC_ADC
```
Comment (MaureenHelm) [inline soc/arm/nxp_lpc/lpc51u68/Kconfig.defconfig.series:38]: This should be `ADC_LPC_ADC`, and added in a later commit after that symbol is introduced

**Review by MaureenHelm — DISMISSED** [review]
Thank you for the contribution!

Can you please take a look at the buildkite failures?

**Comment by Navin-Sankar** [comment]
@dleach02 addressed the conflicts and change requests.
The [hal](https://github.com/zephyrproject-rtos/hal_nxp/pull/95) layer need to be addressed and merged before this pr.

**Comment by Navin-Sankar** [comment]
@dleach02 CI failure is related to this [pull request](https://github.com/zephyrproject-rtos/hal_nxp/pull/113)

**Comment by dleach02** [comment]
> @dleach02 CI failure is related to this [pull request](https://github.com/zephyrproject-rtos/hal_nxp/pull/113)

Yeah, we found out that the LPC and RT6xx had different names for this define. Looking to address this in a patch to the HAL.

**Comment by parthitce** [comment]
Fixes tag in https://github.com/zephyrproject-rtos/hal_nxp/pull/113 closed it. Re-opening.

**Comment by Navin-Sankar** [comment]
@dleach02 Please revisit again

**Comment by dleach02** [comment]
@parthitce, the HAL PR has been merged so you need to resolve the west.yml conflicts

**Comment by dleach02** [comment]
@parthitce, can you rebase this PR and verify the functionality? 

I believe you should be able to just use the west.yml in main now.

**Comment by dleach02** [comment]
@galak, @MaureenHelm can you two confirm if your blocks are still valid on this PR? Thank you.

File: drivers/adc/adc_lpc_adc.c [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+	const struct lpc_adc_config *config = dev->config;
+	struct lpc_adc_data *data = dev->data;
+	ADC_Type *base = config->base;
+	adc_config_t adc_config;
+	uint32_t frequency = 0;
+
+	POWER_DisablePD(kPDRUNCFG_PD_ADC0);
+	POWER_DisablePD(kPDRUNCFG_PD_VD7_ENA);
+	POWER_DisablePD(kPDRUNCFG_PD_VREFP_SW);
+	POWER_DisablePD(kPDRUNCFG_PD_TEMPS);
```
[the remark this answers is not in the document]
Comment (Navin-Sankar) [reply drivers/adc/adc_lpc_adc.c:192]: @mmahadevan108 any updates on this. where to place the adc power related control's?
Either in driver file or soc.c file.

File: dts/arm/nxp/nxp_lpc51u68.dtsi [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+			prescaler = <1>;
+			label = "SC_TIMER";
+			#pwm-cells = <2>;
+		};
+	};
+};
+
+&nvic {
+	arm,num-irq-priority-bits = <2>;
+};
```
Comment (mmahadevan108) [inline dts/arm/nxp/nxp_lpc51u68.dtsi:193]: Please move this dts file to a separate commit.

File: soc/arm/nxp_lpc/lpc51u68/linker.ld [hunk 7]
```diff
@@ -0,0 +1,9 @@
+/*
+ * Copyright (c) 2021 metraTec GmbH
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <autoconf.h>
```
Comment (mmahadevan108) [inline soc/arm/nxp_lpc/lpc51u68/linker.ld:7]: This include can be deleted

File: soc/arm/nxp_lpc/lpc51u68/soc.h [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef _SOC_H_
+#define _SOC_H_
+
+#ifndef _ASMLANGUAGE
+#include <sys/util.h>
+
+#include <devicetree.h>
```
Comment (mmahadevan108) [inline soc/arm/nxp_lpc/lpc51u68/soc.h:13]: This include can be deleted

File: drivers/adc/adc_lpc_adc.c [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+
+	/* Enable Clock for ADC0 channel */
+	CLOCK_EnableClock(kCLOCK_Adc0);
+
+	base->CTRL |= ADC_CTRL_BYPASSCAL_MASK;
+	frequency = CLOCK_GetFreq(kCLOCK_BusClk);
```
Comment (mmahadevan108) [inline drivers/adc/adc_lpc_adc.c:198]: This should be using the `clock_control` driver API. You would need to add device tree properties for the adc. Please refer the LPC UART driver.

File: drivers/adc/adc_lpc_adc.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (mmahadevan108) [reply drivers/adc/adc_lpc_adc.c:192]: Agree with @MaureenHelm, this should not be in the driver. Probably `soc_init` is the place for this

**Review by mmahadevan108 — DISMISSED** [review]

File: drivers/adc/adc_lpc_adc.c (hunk 9 above)
Comment (Navin-Sankar) [reply drivers/adc/adc_lpc_adc.c:198]: I didn't get your point. How to define clock frequency in dts file?

Comment (mmahadevan108) [reply drivers/adc/adc_lpc_adc.c:198]: Below is an example of how we get clock frequency for the I2C driver.
https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/i2c/i2c_mcux_flexcomm.c#L65

Below is an example of how we specify in the device tree which clock
https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/arm/nxp/nxp_lpc55S1x_common.dtsi#L150

Below is the clock control driver for LPC:
https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/clock_control/clock_control_mcux_syscon.c

Hope this helps.

Comment (Navin-Sankar) [reply drivers/adc/adc_lpc_adc.c:198]: Thanks for the hint.

Comment (Navin-Sankar) [reply drivers/adc/adc_lpc_adc.c:198]: @mmahadevan108 the current **clock_control** driver support's only to get the frequency for peripheral clocks like flexcomm, can, timers.
But for this adc driver, **BusClock** is used to get the clock frequency. 

For that we need to introduce a new macro in a header file **dt-bindings/clock/mcux_lpc_syscon_clock.h**  to get the clock frequency of system clock's.

Is this approach is correct?

Comment (mmahadevan108) [reply drivers/adc/adc_lpc_adc.c:198]: Yes, that would be the approach to take.

**Comment by Navin-Sankar** [comment]
@mmahadevan108 Added busclock support [here](https://github.com/zephyrproject-rtos/zephyr/pull/43554)

**Comment by mmahadevan108** [comment]
Would you like to rebase and update

**Comment by dleach02** [comment]
@galak can you review to determine if your change request has been addressed?

File: drivers/adc/adc_lpc_adc.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (mmahadevan108) [reply drivers/adc/adc_lpc_adc.c:195]: @Navin-Sankar, could kindly take a look at addressing this comment.

**Comment by dleach02** [comment]
@MaureenHelm @galak, can you revisit your  change requests to determine if @Navin-Sankar has addressed your items?

File: boards/arm/lpcxpresso51u68/lpcxpresso51u68.dts [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+		red_led: led_2 {
+			gpios = <&gpio0 29 0>;
+			label = "Red LED";
+		};
+	};
+
+	pwmleds {
+		compatible = "pwm-leds";
+		green_pwm_led: green_pwm_led {
+			pwms = <&sc_timer 4 PWM_POLARITY_NORMAL>;
```
Comment (mmahadevan108) [inline boards/arm/lpcxpresso51u68/lpcxpresso51u68.dts:55]: This might need to be updated per changes to PWM. This might be causing the CI failures.

**Review by dleach02 — CHANGES_REQUESTED** [review]
@Navin-Sankar, this will need to be transitioned to using pinctrl

**Comment by dleach02** [comment]
@Navin-Sankar do you plan on revisiting this PR?

**Comment by Navin-Sankar** [comment]
Currently working on this, requested a help from @danieldegrasse to generate the pincontrol file.

**Comment by dleach02** [comment]
@Navin-Sankar you will need to update the west.yml to point to the correct NXP HAL sha that merged your pinctrl changes.

File: soc/arm/nxp_lpc/lpc51u68/Kconfig.defconfig.series [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+config UART_MCUX_FLEXCOMM
+	default y
+	depends on SERIAL
+
+config SOC_FLASH_LPC
+	default y
+	depends on FLASH
+
+config PWM_MCUX_SCTIMER
```
Comment (galak) [inline soc/arm/nxp_lpc/lpc51u68/Kconfig.defconfig.series:34]: remove PWM_MCUX_SCTIMER, not needed as devicetree will enable this.

File: soc/arm/nxp_lpc/lpc51u68/Kconfig.defconfig.series [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+
+config PINMUX_MCUX_LPC
+	default y
+	depends on PINMUX
+
+config GPIO_MCUX_LPC
+	default y
+	depends on GPIO
+
+config UART_MCUX_FLEXCOMM
```
Comment (galak) [inline soc/arm/nxp_lpc/lpc51u68/Kconfig.defconfig.series:26]: remove UART_MCUX_FLEXCOMM as devicetree will enable this.

File: soc/arm/nxp_lpc/lpc51u68/Kconfig.defconfig.series [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+	default 32
+
+config SOC
+	default "lpc51u68"
+
+config PINMUX_MCUX_LPC
+	default y
+	depends on PINMUX
+
+config GPIO_MCUX_LPC
```
Comment (galak) [inline soc/arm/nxp_lpc/lpc51u68/Kconfig.defconfig.series:22]: remove GPIO_MCUX_LPC as devicetree will enable this

File: soc/arm/nxp_lpc/lpc51u68/Kconfig.defconfig.series [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+	default "lpc51u68"
+
+config NUM_IRQS
+	# must be >= the highest interrupt number used.
+	default 32
+
+config SOC
+	default "lpc51u68"
+
+config PINMUX_MCUX_LPC
```
Comment (galak) [inline soc/arm/nxp_lpc/lpc51u68/Kconfig.defconfig.series:18]: Is pinmux needed?

File: drivers/hwinfo/Kconfig [hunk 15]
```diff
@@ -67,7 +67,8 @@ config HWINFO_MCUX_SRC_V2
 config HWINFO_MCUX_SYSCON
 	bool "NXP LPC device ID"
 	default y
-	depends on HAS_MCUX_SYSCON && !SOC_SERIES_LPC54XXX
+	depends on HAS_MCUX_SYSCON && !SOC_SERIES_LPC54XXX \
+		   && !SOC_SERIES_LPC51U68
```
Comment (galak) [inline drivers/hwinfo/Kconfig:71]: Change this to `depends on DT_HAS_NXP_LPC_UID`

**Review by galak — DISMISSED** [review]
Some Kconfig cleanups needed

**Comment by mmahadevan108** [comment]
@Navin-Sankar. Please rebase and remove your changes to `west.yml`

File: dts/arm/nxp/nxp_lpc51u68.dtsi [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+			compatible = "arm,cortex-m0+";
+			reg = <0>;
+		};
+	};
+
+	soc {
+		syscon: syscon@40000000 {
+			compatible = "nxp,lpc-syscon";
+			reg = <0x40000000 0x4000>;
+			label = "SYSCON";
```
Comment (galak) [inline dts/arm/nxp/nxp_lpc51u68.dtsi:28]: remove `label`

File: dts/arm/nxp/nxp_lpc51u68.dtsi [hunk 17]
```diff
@@ -0,0 +1,173 @@
...
+
+		sramx:memory@4000000 {
+			compatible = "mmio-sram";
+			reg = <0x40000000 DT_SIZE_K(32)>;
+		};
+
+		flash0: flash@0 {
+			compatible = "soc-nv-flash";
+			reg = <0 DT_SIZE_K(256)>;
+			label = "LPC_FLASH";
```
[the remark this answers is not in the document]
Comment (Navin-Sankar) [reply dts/arm/nxp/nxp_lpc51u68.dtsi]: done

File: drivers/gpio/gpio_mcux_lpc.c [hunk 18]
```diff
@@ -265,6 +265,9 @@ static uint32_t attach_pin_to_isr(uint32_t port, uint32_t pin, uint32_t isr_no)
 #if (FSL_FEATURE_PINT_NUMBER_OF_CONNECTED_OUTPUTS > 8)
 	#error having more than 8 PINT IRQs not supported in driver
 #elif (FSL_FEATURE_PINT_NUMBER_OF_CONNECTED_OUTPUTS > 4)
+#ifdef CONFIG_SOC_SERIES_LPC51U68
+	pint_idx = isr_no - PIN_INT0_IRQn;
+#endif
```
Comment (mmahadevan108) [inline drivers/gpio/gpio_mcux_lpc.c:270]: Please delete the above 3 lines as LPC51U68 has less than 4 outputs and should never go into this block of code.

**Review by mmahadevan108 — CHANGES_REQUESTED** [review]
Please update the changes to the GPIO driver

**Review by mmahadevan108 — APPROVED** [review]

**Review by dleach02 — APPROVED** [review]
